src/main/scala/viper/silver/ast/utility/rewriter/Strategy.scala

```diff
@@ -400,6 +400,8 @@ class Strategy[N <: Rewritable : reflection.TypeTag : scala.reflect.ClassTag, C
       node match {
         case map: Map[_, _] => map.map(rewriteTopDown(_, context)).asInstanceOf[A]
 
+        case (t1, t2) => (rewriteTopDown(t1, context), rewriteTopDown(t2, context)).asInstanceOf[A]

```

Comment:
Scala has `Tuple2` to `Tuple22`, and we should ideally support them all. Unfortunately, I don't see how to uniformly abstract over constructors of different arity without resorting to reflection, which reduces performance. Do you?

Otherwise, let's take a practical approach here (and in all cases below), and support `Tuples2` to `Tuples5` by copy & paste.

In principle, the copy&paste code could also be generated via a Scala macro (for all tuples), but let's not go down that path now.

Reply:
Your solution uses reflection, which (as mentioned) is possible, but potentially costly in terms of performance. Since we had a significant reflection-induced performance issue a while ago, I'm strongly against risking it again.

I agree that supporting only a few tuples isn't ideal - and why stop at `Tuple5`, and not at `Tuple7`? But then again, so is Scala's choice to stop at `Tuple22` (why not 25?). Here are my arguments why it's nonetheless OK as a practical solution:

  1. The Silver AST transformer has many users beyond the Silver code base (e.g. frontends), which may not stop at `Tuple2`.
 
  2. On the other hand, large tuples are bad design anyway, and should be replaced by a (case) class. Smaller tuples - pairs, triples, maybe also quadruples, maybe even quintuples - are at least not uncommon, so we should support them.

I'm therefore against supporting only `Tuple2`. I'd be OK with supporting up to `Tuple5`, but feel free to also support 6, 7, or all 22.

Reply:
I see. With commit 741e6d416dcfd054a07a4d1dd1a49d2429a2d976 `Tuple2` to `Tuple5` is now supported with a copy&paste approach. Firstly, I tried to implement a more general solution. Something like

```
object Tuple {
  def apply[T1,T2](t1:T1, t2:T2): (T1,T2) = (t1,t2)
  def apply[T1,T2,T3](t1:T1, t2:T2, t3: T3): (T1,T2,T3) = (t1,t2,t3)
  ...
  
  
  def unapply[T1,T2](t: (T1,T2)): Option[(T1, T2)] = Some(t)
  def unapply[T1,T2,T3](t: (T1,T2,T3)): Option[(T1, T2, T3)] = Some(t)
  ... 
  def unapply[A](x: A): Option[Nothing] = None
  

  def map[A,B](x: (A,A), f: A => B): (B,B) = (f(x._1), f(x._2))
  ...
  def map[A,B](x: A, f: B) = throw new Error(s"Tuple: Either 'x' was not a tuple or function 'f' does not match the Tuple")
}

```
but I struggled to get the type interference working. :(